src/Stratis.Bitcoin.Features.BlockStore/BlockStoreQueue.cs

```diff
@@ -271,89 +275,77 @@ public void AddToPending(ChainedHeaderBlock chainedHeaderBlock)
-                // or the max batch size is reached or the node is shutting down.
-                if (dequeueTask.Status == TaskStatus.RanToCompletion)
-                {
-                    ChainedHeaderBlock item = dequeueTask.Result;
-
-                    // Set the dequeue task to null so it can be assigned on the next iteration.
-                    dequeueTask = null;
-
-                    lock (this.getBlockLock)
+                    using (CancellationTokenSource timeoutTokenSource = CancellationTokenSource.CreateLinkedTokenSource(this.nodeLifetime.ApplicationStopping))
                     {
-                        this.batch.Add(item);
-                    }
+                        timeoutTokenSource.CancelAfter(BatchMaxSaveIntervalSeconds * 1000);
+                        await this.addEvent.WaitAsync(timeoutTokenSource.Token).ConfigureAwait(false);
 
-                    this.currentBatchSizeBytes += item.Block.BlockSize.Value;
-
-                    saveBatch = saveBatch || (this.currentBatchSizeBytes >= BatchThresholdSizeBytes) || this.chainState.IsAtBestChainTip;
+                        // The event was set, indicating at least one item is in the batch.
+                        // We check further down whether this was sufficient to trigger a save.
+                        // Reset the event for now.
+                        this.addEvent.Reset();
+                    }
                 }
-                else
+                catch (OperationCanceledException)
                 {
-                    // Will be executed in case timer ran out or node is being shut down.
+                    // Check whether it was a node shutdown/Dispose() or the timeout token.
+                    if (this.nodeLifetime.ApplicationStopping.IsCancellationRequested)
+                        this.logger.LogDebug("Node is shutting down. Save batch.");
+                    else
+                        this.logger.LogDebug("Maximum batch saving interval elapsed. Save batch.");
+
+                    // When node is shutting down or Dispose() is called,
+                    // we want to save whatever is in the batch before exiting the loop.
+                    // For normal timeouts we just save the batch, the loop will execute again.
                     saveBatch = true;
                 }
 
+                // Apart from node shutdown, we also trigger a save if the batch is sufficiently full,
+                // or we are at the consensus tip.
+                saveBatch = saveBatch || (this.currentBatchSizeBytes >= BatchThresholdSizeBytes) || this.chainState.IsAtBestChainTip;
+
                 if (saveBatch)
                 {
                     if (this.batch.Count != 0)
                     {
-                        await this.SaveBatchAsync().ConfigureAwait(false);
+                        List<ChainedHeaderBlock> saved = await this.SaveBatchAsync().ConfigureAwait(false);
```

Comment:
`this.batch.Count` should probably be inside the lock no?

Reply:
Perhaps. But all we really care about is that the count is nonzero. Since this is the only thread that actually removes items from the batch, it should be safe.